File: src/com/douma/written_test/huawei/_20210831/_1_MaxFav.py
# 抖码算法，让算法学习变得简单有趣
# 作者：菲菲

if __name__ == '__main__':
    data = str(input()).split(" ")
    x, n = int(data[0]), int(data[1])
    w, p, v = [0] * n, [0] * n, [0] * n
    for i in range(n):
        data = str(input()).split(" ")
        w[i], p[i], v[i] = int(data[0]), int(data[1]), int(data[2])

    # 把多重背包问题，转化成 0-1 背包问题
    # 比如，有 2 件价值为 5，重量为 2 的同一物品，
    # 我们就可以分为物品 a和物品 b，a 和 b 的价值都为 5，重量都为 2，
    # 但我们把它们视作不同的物品。
    new_n = sum(p)
    new_w, new_v, index = [0] * new_n, [0] * new_n, 0
    for i in range(n):
        for j in range(p[i]):
            new_w[index], new_v[index], index = w[i], v[i], index + 1

    # 1. 状态定义：dp[c] : 将物品放入容量为 c 的背包中产生的最大价值
    dp = [0] * (x + 1)

    # 2. 状态初始化

    # 3. 状态转移
    for i in range(new_n):
        for c in range(x, new_w[i] - 1, -1):
            dp[c] = max(dp[c], new_v[i] + dp[c - new_w[i]])

    print(dp[x])

File: src/com/douma/笔试代码/huawei/_20210831/_1_MaxFav.java
package com.douma.笔试代码.huawei._20210831;

import java.util.Scanner;

/**
 * 抖码算法，让算法学习变的简单有趣
 *
 * @作者 : 菲菲
 */
public class _1_MaxFav {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int x = scanner.nextInt(); // 可用的钱总额
        int n = scanner.nextInt(); // 商品的种类
        int[] w = new int[n]; // 每种商品的价格
        int[] p = new int[n]; // 每种商品的数量
        int[] v = new int[n]; // 每种商品的喜爱度
        for (int i = 0; i < n; i++) {
            w[i] = scanner.nextInt();
            p[i] = scanner.nextInt();
            v[i] = scanner.nextInt();
        }

        // 把多重背包问题，转化成 0-1 背包问题
        // 比如，有 2 件价值为 5，重量为 2 的同一物品，
        // 我们就可以分为物品 a和物品 b，a 和 b 的价值都为 5，重量都为 2，
        // 但我们把它们视作不同的物品。
        int newN = 0;
        for (int i = 0; i < n; i++) {
            newN += p[i];
        }

        int[] newW = new int[newN];
        int[] newV = new int[newN];
        int index = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < p[i]; j++) {
                newW[index] = w[i];
                newV[index] = v[i];
                index++;
            }
        }

        // 1. 状态定义：dp[c] : 将物品放入容量为 c 的背包中产生的最大价值
        int[] dp = new int[x + 1];

        // 2. 状态初始化

        // 3. 状态转移
        for (int i = 0; i < newN; i++) {
            for (int c = x; c >= newW[i]; c--) {
                dp[c] = Math.max(dp[c], newV[i] + dp[c - newW[i]]);
            }
        }

        // 4. 返回结果
        System.out.println(dp[x]);
    }

}
